pyquil/operator_estimation.py

```diff
@@ -0,0 +1,428 @@
+                'qubits': o.qubits,
+            }
+        if isinstance(o, ExperimentResult):
+            return {
+                'type': 'ExperimentResult',
+                'experiment': o.experiment,
+                'expectation': o.expectation,
+                'stddev': o.stddev,
+            }
+
+        return o
+
+
+def to_json(fn, obj):
+    with open(fn, 'w') as f:
+        json.dump(obj, f, cls=OperatorEncoder, indent=2)
+    return fn
+
+
+def _operator_object_hook(obj):
+    if 'type' in obj and obj['type'] == 'ExperimentSuite':
+        return ExperimentSuite([[Experiment.from_str(e) for e in exps]
+                                for exps in obj['experiments']],
+                               program=Program(obj['program']),
+                               qubits=obj['qubits'])
+    return obj
+
+
+def read_json(fn):
+    with open(fn) as f:
+        return json.load(f, object_hook=_operator_object_hook)
+
+
+def _local_pauli_eig_prep(op: str, idx: int):
+    """
+    Generate gate sequence to prepare a the +1 eigenstate of a Pauli operator, assuming
+    we are starting from the ground state (the +1 eigenstate of Z^{\\otimes n}).
+
+    :param op: A string representation of the Pauli operator whose eigenstate we'd like to prepare.
+    :param idx: The index of the qubit that the preparation is acting on
+    :return: A program which will prepare the requested state.
+    """
+    if op == 'X':
+        gate = RY(pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(-pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+
+    return Program(gate)
+
+
+def _local_pauli_eig_meas(op, idx):
+    """
+    Generate gate sequence to measure in the eigenbasis of a Pauli operator, assuming
+    we are only able to measure in the Z eigenbasis.
+
+    """
+    if op == 'X':
+        gate = RY(-pi / 2, idx)
+    elif op == 'Y':
+        gate = RX(pi / 2, idx)
+    elif op == 'Z':
+        return Program()
+    else:
+        raise ValueError('Unknown gate operation')
+    return Program(gate)
+
+
+def _ops_belong_to_a_tpb(op_code1: str, op_code2: str):
+    """
+    Given two op strings (I, X, Y, or Z) return whether they share a tensor product basis.
+
+    I.e. are they the same or is one of them 'I'.
+    """
+    if op_code1 == op_code2:
+        # Same op
+        return True
+    elif op_code1 == 'I' or op_code2 == 'I':
+        # I commutes with everything
+        return True
+    else:
+        # Otherwise, they do not commute.
+        return False
+
+
+def _all_qubits_belong_to_a_tpb(op1: PauliTerm, op2: PauliTerm):
+    """
+    Compare all qubits between two PauliTerms to see if they share an overall tensor product basis.
+    """
+    all_qubits = set(op1.get_qubits()) | set(op2.get_qubits())
+    return all(_ops_belong_to_a_tpb(op1[q], op2[q]) for q in all_qubits)
+
+
+def construct_tpb_graph(experiments: ExperimentSuite):
```

Comment:
This is an issue related to the comment about the `append` function above. The `construct_tpb_graph(..)` function will raise `TypeError: object of type 'Experiment' has no len()` if fed in an ExperimentSuite that also has an Experiment in it (and not a list of Experiment). For example, if we did
```
import pyquil.operator_estimation as oe
import pyquil.paulis as pl
expt1 = oe.Experiment(pl.sX(1) * pl.sZ(0), pl.sX(2) * pl.sY(1) * pl.sZ(0))
expt2 = oe.Experiment(pl.sZ(1) * pl.sX(0), pl.sY(2) * pl.sZ(1) * pl.sX(0))
expt3 = oe.Experiment(pl.sX(2) * pl.sY(1) * pl.sZ(0), pl.sZ(2) * pl.sX(1) * pl.sY(0))
expt_list = [expt1, expt2, expt3]
expt_suite = oe.ExperimentSuite(expt_list, Program(H(0), H(1), H(2)), [0, 1, 2])
expt_suite.append(Experiment(pl.sX(0), pl.sZ(1)))
```
and then do
```
oe.construct_tpb_graph(expt_suite)
```
we'd get the error msg: `TypeError: object of type 'Experiment' has no len()`.

Reply:
fixed by fixing append.